--- "./Project_FARSI_orig/specs/database_data/parsing/inputs/edge_detection_2_database - Task Itr Count.csv"
+++ "./Project_FARSI/specs/database_data/parsing/inputs/edge_detection_2_database - Task Itr Count.csv"
@@ -0,0 +1,9 @@
+Task Name,number of iterations
+souurce,1
+gaussianSmoothing_2,1638400
+laplacianEstimate_2,1638400
+computeZeroCrossings_2,1638400
+computeGradient_2,1638400
+computeMaxGradientLeaf_2,256
+rejectZeroCrossings_2,1638400
+siink,1
