--- "./Project_FARSI_orig/specs/database_data/parsing/inputs/SOC_example_8p_database - Task Itr Count.csv"
+++ "./Project_FARSI/specs/database_data/parsing/inputs/SOC_example_8p_database - Task Itr Count.csv"
@@ -0,0 +1,11 @@
+Task Name,number of iterations
+souurce,1
+participant0,1000000
+participant1,1000000
+participant2,1000000
+participant3,1000000
+participant4,1000000
+participant5,1000000
+participant6,1000000
+participant7,1000000
+siink,1
