--- "./Project_FARSI_orig/specs/database_data/parsing/inputs/SOC_example_8p_database - Task PE Area.csv"
+++ "./Project_FARSI/specs/database_data/parsing/inputs/SOC_example_8p_database - Task PE Area.csv"
@@ -0,0 +1,11 @@
+Task Name,A53
+souurce,0
+participant0,10
+participant1,10
+participant2,10
+participant3,10
+participant4,10
+participant5,10
+participant6,10
+participant7,10
+siink,0
